<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
    <head>
        <title></title>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>

        <h1 align="left" style="width: 60%;"">Workshop:&nbsp;Introduction into high-throughput data analysis
            using MapMan and Robin</h1>
        <div style="width: 50%; " align="justify">
            <h1 style="margin-top: 1em; margin-right: 0px; margin-bottom: 1em; margin-left: 0px; padding-top: 0px; padding-right: 0px; padding-bottom: 0px; padding-left: 0px; font-size: 1.8em; font-weight: bold; "><span style="font-size: medium; "><span style="color: rgb(128, 128, 128); ">10-11<sup>th </sup>May 2010 in Potsdam-Golm</span></span></h1>

            <h3>Introduction</h3>
            <p>High throughput experiments like microarray-based expression studies, large-scale proteomics
                and metabolimics generate enormous amounts of data that is usually too confusing to be directly
                viewed or interpreted by the experimentator. The data has to be passed through several preprocessing
                steps in order to evaluate the quality, eliminate technical bias and statistically assess the
                data. Robin is a user-friendly graphical application that was specifically tailored to aid the
                individual researcher in the task of microarray data quality assessment, normalization and
                statistical inference of differential gene expression. Using Robin, Affymetrix, generic single
                channel (e.g. AGILENT) and two color microarrays can be processed in an self-explanatory and
                intuitive way, yielding overview of the data quality, experiment results and statistical analysis.
                <br /><br>
                The results generated by Robin are formatted for direct import into the MapMan application.
                Using a specialized partly hand-curated functional ontology that classifies genes in so-called
                &quot;Bins&quot;, MapMan offers a feature-rich environment to view and analyse the data in its biological
                context. MapMan is not restricted to transcript data generated in microarray analyses but can
                also be used with proteomics and metabolomics data and is even able to analyse all three data
                types together. In addition to visualizing differential expression levels on biological pathways,
                MapMan provides many useful functions like time-course display, clustering, Venn diagrams, statistical
                analysis of system responses etc.</p>
            <h3>Workshop</h3>
            <p>The workshop will be held at the Max-Planck-Institute of molecular plant physiology in <a href="http://maps.google.de/maps?f=q&amp;source=s_q&amp;hl=de&amp;geocode=&amp;q=golm,+am+m%C3%BChlenberg+1&amp;sll=52.406463,12.966904&amp;sspn=0.01906,0.04416&amp;ie=UTF8&amp;hq=&amp;hnear=Am+M%C3%BChlenberg+1,+14476+Potsdam,+Brandenburg&amp;z=16&amp;iwloc=A">Golm</a>
                on the 10th and 11th of May 2010. Please find the <a href="http://mapman.gabipd.org/web/guest/workshop-program">program here</a>. Workshop participation is free,
                but the participants will&nbsp;have to cover their travel and accomodation expenses. Up to 18 workstations
                are available -&nbsp;participants who would like to bring their own laptops are invited to do so but are asked to
                indicate that in their registration and install MapMan and Robin beforehand. To make sure that the applications
                will run smoothly, the laptops should have a screen resolution of at least 1280x800, 2GB RAM and
                a modern processor (e.g. Core 2 Duo or comparable).&nbsp;Although we can only provide 18 workstations
                we accept up to 24 participants provided that enough people bring their own laptops. To register, simply
                send an email stating your interest, affiliation and scientific background to <a href="mailto:lohse@mpimp-golm.mpg.de?subject=Workshop_may2010">Marc Lohse</a>.
                Data sets for demonstration of the  applications will be provided at the workshop, nevertheless, participants are invited
                to bring their own data.
                <br />
                The first day will focus on microarray preprocessing and statistical analysis of differential
                expression giving a general introduction followed by hands-on demonstrations of the Robin application.
                On the second day, MapMan will be introduced and the various tools offered by this environment
                will be employed to get a deeper insight into the biology hidden in the data.</p>
            <h3>Registration</h3>
            <p>Please send an expression of interest to participate via email to <a href="mailto:lohse@mpimp-golm.mpg.de?subject=Workshop_may2010">Marc Lohse</a> also stating how
                many people from your lab would like to attend. To make sure that as many individual labs as possible
                can benefit from the workshop, we may limit attendance to one person per lab if the number
                of interested people exceeds the workshop capacity.<br><br>
                Please visit the institute's website for detailed <a href="http://www-en.mpimp-golm.mpg.de/09-travelinfo/index.html">travel information</a>.<br><br>
                News, announcements and workshop material will be made available via this
                webpage so please check back to stay up to date. People having expressed interest via email will
                be informed automatically.</p>            
        </div>


    </body>
</html>
