<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Gene Set Test</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<link rel="stylesheet" type="text/css" href="../../R.css">
</head><body>

<table width="100%" summary="page for geneSetTest {limma}"><tr><td>geneSetTest {limma}</td><td align="right">R Documentation</td></tr></table>
<h2>Gene Set Test</h2>


<h3>Description</h3>

<p>
Test whether a set of genes is enriched for differential expression.
Genes are assumed to be independent.
</p>


<h3>Usage</h3>

<pre>
geneSetTest(selected,statistics,alternative="mixed",type="auto",ranks.only=TRUE,nsim=10000)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>selected</code></td>
<td>
vector specifying the elements of <code>statistic</code> in the test group.  This can be a vector of indices, or a logical vector of the same length as <code>statistics</code>, or any vector such as <code>statistic[selected]</code> contains the statistic values for the selected group.</td></tr>
<tr valign="top"><td><code>statistics</code></td>
<td>
numeric vector giving the values of the test statistic for every gene or probe in the reference set, usually every probe on the microarray.</td></tr>
<tr valign="top"><td><code>alternative</code></td>
<td>
character string specifying the alternative hypothesis, must be one of <code>"mixed"</code> (default), <code>"either"</code>, <code>"up"</code> or <code>"down"</code>. <code>two.sided</code>, <code>"greater"</code> and <code>"less"</code> are also permitted as synonyms for <code>"either"</code>, <code>"up"</code> and <code>"down"</code> respectively.</td></tr>
<tr valign="top"><td><code>type</code></td>
<td>
character string specifying whether the statistics are t-like (<code>"t"</code>), F-like <code>"f"</code> or whether the function should make an educated guess (<code>"auto"</code>)</td></tr>
<tr valign="top"><td><code>ranks.only</code></td>
<td>
logical, if <code>TRUE</code> only the ranks of the <code>statistics</code> are used.</td></tr>
<tr valign="top"><td><code>nsim</code></td>
<td>
number of random samples to take in computing the p-value.
Not used if <code>ranks.only=TRUE</code>.</td></tr>
</table>

<h3>Details</h3>

<p>
This function computes a p-value to test the hypothesis that the selected set of genes tends to be more highly ranked in differential expression compared to randomly selected genes.
This function can be used to detect differential expression for a group of genes, even when the effects are too small or there is too little data to detect the genes individually.
It is also provides a means to compare the results between different experiments.
</p>
<p>
The aim here is the same as for Gene Set Enrichment Analysis (GSEA) introduced by Mootha et al (2003) and Subramanian et al (2005), but the statistical tests used are different.
There are two key differences.
Firstly, <code>geneSetTest</code> assumes that the different genes (or probes) are independent.
(Strictly speaking, it assumes that the genes in the set are no more correlated on average than randomly selected genes.)
This assumption may be reasonable if the gene set is relatively small and if there is relatively little genotypic variation in the data, for example if the data is obtained from genetically identical inbred mice.
The independence assumption may be misleading if the gene set is large or if the data contains a lot of genotypic variation, for example for human cancer samples.
Secondly, the <code>geneSetTest</code> is purely competitive in that the gene set is compared to other genes, rather than a self-contained test of differential expressed for the genes in the gene set (Goeman and Buhlmann, 2007).
These assumptions, when valid, permit a much quicker and more powerful significance test to be conducted.
</p>
<p>
The <code>statistics</code> are usually a set of probe-wise statistics arising for some comparison from a microarray experiment.
They may be t-statistics, meaning that the genewise null hypotheses would be rejected for large positive or negative values, or they may be F-statistics, meaning that only large values are significant.
Any set of signed statistics, such as log-ratios, M-values or moderated t-statistics, are treated as t-like.
Any set of unsigned statistics, such as F-statistics, posterior probabilities or chi-square tests are treated as F-like.
If <code>type="auto"</code> then the statistics will be taken to be t-like if they take both positive and negative values and otherwise will be taken to be F-like.
</p>
<p>
There are four possible alternatives to test for.
<code>alternative=="up"</code> means the genes in the set tend to be up-regulated, with positive t-statistics.
<code>alternative=="down"</code> means the genes in the set tend to be down-regulated, with negative t-statistics.
<code>alternative=="either"</code> means the set is either up or down-regulated as a whole.
<code>alternative=="mixed"</code> test whether the genes in the set tend to be differentially expressed, without regard for direction.
In this case, the test will be significant if the set contains mostly large test statistics, even if some are positive and some are negative.
</p>
<p>
The first three alternativea appropriate if you have a prior expection that all the genes in the set will react in the same direction.
The <code>"mixed"</code> alternative is appropriate if you know only that the genes are involved in the relevant pathways, without knowing the direction of effect for each gene.
The <code>"mixed"</code> alternative is the only one possible with F-like statistics.
</p>
<p>
The test statistic used for the gene-set-test is the mean of the statistics in the set.
If <code>ranks.only</code> is <code>TRUE</code> the only the ranks of the statistics are used.
In this case the p-value is obtained from a Wilcoxon test.
If <code>ranks.only</code> is <code>FALSE</code>, then the p-value is obtained by simulation using <code>nsim</code> random selected sets of genes.
</p>


<h3>Value</h3>

<p>
Numeric value giving the estimated p-value.</p>

<h3>Author(s)</h3>

<p>
Gordon Smyth
</p>


<h3>References</h3>

<p>
Goeman, JJ, and Buhlmann P 
(2007)
Analyzing gene expression data in terms of gene sets: methodological issues.
<EM>Bioinformatics</EM> 23, 980-987. 
</p>
<p>
Mootha, V. K.,
Lindgren, C. M.,
Eriksson, K. F.,
Subramanian, A.,
Sihag, S.,
Lehar, J.,
Puigserver, P.,
Carlsson, E.,
Ridderstrale, M.,
Laurila, E.,
Houstis, N.,
Daly, M. J.,
Patterson, N.,
Mesirov, J. P.,
Golub, T. R.,
Tamayo, P.,
Spiegelman, B.,
Lander, E. S.,
Hirschhorn, J. N.,
Altshuler, D.,
Groop, L. C.
(2003). 
PGC-1alpha-responsive genes involved in oxidative phosphorylation are coordinately downregulated in human diabetes.
<EM>Nature Genetics</EM> 34, 267-273.
</p>
<p>
Subramanian A, Tamayo P, Mootha VK, Mukherjee S, Ebert BL, Gillette MA, Paulovich A, Pomeroy SL, Golub TR, Lander ES, Mesirov JP.
(2005).
Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles.
<EM>Proc Natl Acad Sci USA</EM> 102, 15545-15550.
</p>


<h3>See Also</h3>

<p>
<code><a href="../../stats/html/wilcox.test.html">wilcox.test</a></code>
</p>


<h3>Examples</h3>

<pre>
stat &lt;- -9:9
sel &lt;- c(2,4,5)
geneSetTest(sel,stat,alternative="down")
geneSetTest(sel,stat,alternative="either")
geneSetTest(sel,stat,alternative="down",ranks=FALSE)
sel &lt;- c(1,19)
geneSetTest(sel,stat,alternative="mixed")
geneSetTest(sel,stat,alternative="mixed",ranks=FALSE)
</pre>



<hr><div align="center">[Package <em>limma</em> version 2.14.7 <a href="00Index.html">Index]</a></div>

</body></html>
