<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Output Significant Genes</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<link rel="stylesheet" type="text/css" href="../../R.css">
</head><body>

<table width="100%" summary="page for topGene {RankProd}"><tr><td>topGene {RankProd}</td><td align="right">R Documentation</td></tr></table>
<h2>Output Significant Genes</h2>


<h3>Description</h3>

<p>
Identify differentially expressed genes using 
rank product method
</p>


<h3>Usage</h3>

<pre>

     topGene(x,cutoff=NULL,method="pfp",num.gene=NULL,logged=TRUE,logbase=2,gene.names=NULL)</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
the value returned by the 
function RP, RP.advance or
Rsum.advance </td></tr>
<tr valign="top"><td><code>cutoff</code></td>
<td>
threshold in pfp used to select genes</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>
</td></tr>
<tr valign="top"><td><code>num.gene</code></td>
<td>
number of candidate genes of interests, 
if cutoff is provided, this will be ignored</td></tr>
<tr valign="top"><td><code>logged</code></td>
<td>
if "TRUE", data has bee logged, otherwise set it 
to "FALSE"</td></tr>
<tr valign="top"><td><code>logbase</code></td>
<td>
base used when taking log, used to restore the 
fold change.The default value is 2, this will be 
ignored if logged=FALSE</td></tr>
<tr valign="top"><td><code>gene.names</code></td>
<td>
if "NULL", no gene name will be 
attached to the output table</td></tr>
</table>

<h3>Value</h3>

<p>
Two tables of identified genes with 
gene.index: index of gene in the original data set 
RP/Rsum: Computed rank product/sum for each gene
FC:(class1/class2): Expression Fold change of class 1/ class 2.                   
pfp: estimated pfp for each gene if the gene is used as cutoff point
P.value: estimated p-value for each gene 
<br>
Table 1 list genes that are up-regulated under class 2, Table 1 ist 
genes that are down-regulated under class 2,</p>

<h3>Author(s)</h3>

<p>
Fangxin Hong <a href="mailto:fhong@salk.edu">fhong@salk.edu</a>
</p>


<h3>References</h3>

<p>
Breitling, R., Armengaud, P., Amtmann, A., and Herzyk, 
P.(2004) Rank Products: A simple, yet powerful, new method 
to detect differentially regulated genes in
replicated microarray experiments, <EM>FEBS Letter</EM>, 57383-92
</p>


<h3>See Also</h3>

<p>
<code><a href="plotRP.html">plotRP</a></code> <code><a href="RP.html">RP</a></code>  
<code><a href="RPadvance.html">RPadvance</a></code> <code><a href="RSadvance.html">RSadvance</a></code>
</p>


<h3>Examples</h3>

<pre>

      # Load the data of Golub et al. (1999). data(golub) 
      # contains a 3051x38 gene expression
      # matrix called golub, a vector of length called golub.cl 
      # that consists of the 38 class labels,
      # and a matrix called golub.gnames whose third column 
      # contains the gene names.
      data(golub)

      #use a subset of data as example, apply the rank 
      #product method
      subset &lt;- c(1:4,28:30)
      #Setting rand=123, to make the results reproducible,

      #identify genes 
      RP.out &lt;- RP(golub[,subset],golub.cl[subset],rand=123)  

      #get two lists of differentially expressed genes 
      #by setting FDR (false discivery rate) =0.05

      table=topGene(RP.out,cutoff=0.05,method="pfp",logged=TRUE,logbase=2,
                   gene.names=golub.gnames[,3])
      table$Table1
      table$Table2

      #using pvalue&lt;0.05
      topGene(RP.out,cutoff=0.05,method="pval",logged=TRUE,logbase=2,
                   gene.names=golub.gnames[,3])

      #by selecting top 10 genes

      topGene(RP.out,num.gene=10,gene.names=golub.gnames[,3])

</pre>



<hr><div align="center">[Package <em>RankProd</em> version 2.12.0 <a href="00Index.html">Index]</a></div>

</body></html>
