/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

/*
 * RNASeqBAMSAMImportDialog.java
 *
 * Created on 10.05.2012, 14:35:03
 */
package de.mpimp.golm.robin.rnaseq.mapping;

import de.mpimp.golm.common.gui.SimpleErrorMessage;
import de.mpimp.golm.common.help.HelpHandler;
import de.mpimp.golm.robin.GUI.RobinMainGUI;
import de.mpimp.golm.robin.data.RNASeqDataModel;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.FileDialog;
import java.awt.Frame;
import java.awt.Point;
import java.io.File;
import net.sf.samtools.SAMFileReader;

/**
 *
 * @author marc
 */
public class RNASeqBAMSAMImportDialog extends javax.swing.JDialog {

    private RobinMainGUI main;
    private RNASeqDataModel dataModel;
    private boolean cancelled = false;

    /** Creates new form RNASeqBAMSAMImportDialog */
    public RNASeqBAMSAMImportDialog(RobinMainGUI parent, boolean modal, RNASeqDataModel dataModel) {
        super(parent, modal);
        this.main = parent;
        this.dataModel = dataModel;
        initComponents();
        centerOnParent();
    }

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {
        bindingGroup = new org.jdesktop.beansbinding.BindingGroup();

        refTypeButtonGroup = new javax.swing.ButtonGroup();
        jLabel1 = new javax.swing.JLabel();
        transcriptomeButton = new javax.swing.JRadioButton();
        genomeButton = new javax.swing.JRadioButton();
        loadGFFButton = new javax.swing.JButton();
        importButton = new javax.swing.JButton();
        cancelButton = new javax.swing.JButton();
        annoFileField = new javax.swing.JTextField();
        bamsamImportHelp = new javax.swing.JButton();
        jLabel2 = new javax.swing.JLabel();
        validationStringencyBox = new javax.swing.JComboBox();
        bamsamImportStringencyHelp = new javax.swing.JButton();

        setDefaultCloseOperation(javax.swing.WindowConstants.DISPOSE_ON_CLOSE);

        jLabel1.setFont(jLabel1.getFont().deriveFont(jLabel1.getFont().getStyle() | java.awt.Font.BOLD, jLabel1.getFont().getSize()+1));
        jLabel1.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.jLabel1.text")); // NOI18N

        refTypeButtonGroup.add(transcriptomeButton);
        transcriptomeButton.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.transcriptomeButton.text")); // NOI18N
        transcriptomeButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                transcriptomeButtonActionPerformed(evt);
            }
        });

        refTypeButtonGroup.add(genomeButton);
        genomeButton.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.genomeButton.text")); // NOI18N
        genomeButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                genomeButtonActionPerformed(evt);
            }
        });

        loadGFFButton.setIcon(new javax.swing.ImageIcon(getClass().getResource("/de/mpimp/golm/robin/resources/openFile.png"))); // NOI18N
        loadGFFButton.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.loadGFFButton.text")); // NOI18N

        org.jdesktop.beansbinding.Binding binding = org.jdesktop.beansbinding.Bindings.createAutoBinding(org.jdesktop.beansbinding.AutoBinding.UpdateStrategy.READ_WRITE, genomeButton, org.jdesktop.beansbinding.ELProperty.create("${selected}"), loadGFFButton, org.jdesktop.beansbinding.BeanProperty.create("enabled"));
        bindingGroup.addBinding(binding);

        loadGFFButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadGFFButtonActionPerformed(evt);
            }
        });

        importButton.setIcon(new javax.swing.ImageIcon(getClass().getResource("/de/mpimp/golm/robin/resources/GoToNextHS.png"))); // NOI18N
        importButton.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.importButton.text")); // NOI18N
        importButton.setEnabled(false);

        binding = org.jdesktop.beansbinding.Bindings.createAutoBinding(org.jdesktop.beansbinding.AutoBinding.UpdateStrategy.READ_WRITE, refTypeButtonGroup, org.jdesktop.beansbinding.ELProperty.create("${selection.selected}"), importButton, org.jdesktop.beansbinding.BeanProperty.create("enabled"));
        bindingGroup.addBinding(binding);

        importButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                importButtonActionPerformed(evt);
            }
        });

        cancelButton.setIcon(new javax.swing.ImageIcon(getClass().getResource("/de/mpimp/golm/robin/resources/close_icon.png"))); // NOI18N
        cancelButton.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.cancelButton.text")); // NOI18N
        cancelButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                cancelButtonActionPerformed(evt);
            }
        });

        annoFileField.setEditable(false);
        annoFileField.setFont(annoFileField.getFont().deriveFont(annoFileField.getFont().getSize()-3f));
        annoFileField.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.annoFileField.text")); // NOI18N

        binding = org.jdesktop.beansbinding.Bindings.createAutoBinding(org.jdesktop.beansbinding.AutoBinding.UpdateStrategy.READ_WRITE, genomeButton, org.jdesktop.beansbinding.ELProperty.create("${selected}"), annoFileField, org.jdesktop.beansbinding.BeanProperty.create("enabled"));
        bindingGroup.addBinding(binding);

        bamsamImportHelp.setIcon(new javax.swing.ImageIcon(getClass().getResource("/de/mpimp/golm/robin/resources/QuestionMark.png"))); // NOI18N
        bamsamImportHelp.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.bamsamImportHelp.text")); // NOI18N
        bamsamImportHelp.setActionCommand(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.bamsamImportHelp.actionCommand")); // NOI18N
        bamsamImportHelp.setBorder(null);
        bamsamImportHelp.setBorderPainted(false);
        bamsamImportHelp.setContentAreaFilled(false);
        bamsamImportHelp.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                bamsamImportHelpActionPerformed(evt);
            }
        });

        jLabel2.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.jLabel2.text")); // NOI18N

        validationStringencyBox.setModel(new javax.swing.DefaultComboBoxModel(new String[] { "strict", "lenient", "silent" }));

        bamsamImportStringencyHelp.setIcon(new javax.swing.ImageIcon(getClass().getResource("/de/mpimp/golm/robin/resources/QuestionMark.png"))); // NOI18N
        bamsamImportStringencyHelp.setText(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.bamsamImportStringencyHelp.text")); // NOI18N
        bamsamImportStringencyHelp.setActionCommand(org.openide.util.NbBundle.getMessage(RNASeqBAMSAMImportDialog.class, "RNASeqBAMSAMImportDialog.bamsamImportStringencyHelp.actionCommand")); // NOI18N
        bamsamImportStringencyHelp.setBorder(null);
        bamsamImportStringencyHelp.setBorderPainted(false);
        bamsamImportStringencyHelp.setContentAreaFilled(false);
        bamsamImportStringencyHelp.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                bamsamImportStringencyHelpActionPerformed(evt);
            }
        });

        org.jdesktop.layout.GroupLayout layout = new org.jdesktop.layout.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(layout.createSequentialGroup()
                .addContainerGap()
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                    .add(layout.createSequentialGroup()
                        .add(transcriptomeButton, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, 191, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(org.jdesktop.layout.LayoutStyle.UNRELATED)
                        .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                            .add(layout.createSequentialGroup()
                                .add(29, 29, 29)
                                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.TRAILING, false)
                                    .add(org.jdesktop.layout.GroupLayout.LEADING, annoFileField)
                                    .add(org.jdesktop.layout.GroupLayout.LEADING, loadGFFButton))
                                .addContainerGap())
                            .add(org.jdesktop.layout.GroupLayout.TRAILING, layout.createSequentialGroup()
                                .add(genomeButton, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                                .add(29, 29, 29))))
                    .add(layout.createSequentialGroup()
                        .add(jLabel1, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                        .add(bamsamImportHelp)
                        .addContainerGap())))
            .add(layout.createSequentialGroup()
                .add(18, 18, 18)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                    .add(layout.createSequentialGroup()
                        .add(cancelButton)
                        .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                        .add(importButton))
                    .add(layout.createSequentialGroup()
                        .add(jLabel2, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                        .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                        .add(validationStringencyBox, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, 99, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                        .add(bamsamImportStringencyHelp, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                        .add(32, 32, 32)))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(layout.createSequentialGroup()
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                    .add(jLabel1, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                    .add(layout.createSequentialGroup()
                        .addContainerGap()
                        .add(bamsamImportHelp)))
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.BASELINE)
                    .add(transcriptomeButton)
                    .add(genomeButton))
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(loadGFFButton)
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(annoFileField, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED, 27, Short.MAX_VALUE)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.CENTER)
                    .add(jLabel2)
                    .add(validationStringencyBox, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                    .add(bamsamImportStringencyHelp))
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.UNRELATED)
                .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.BASELINE)
                    .add(cancelButton)
                    .add(importButton)))
        );

        bindingGroup.bind();

        pack();
    }// </editor-fold>//GEN-END:initComponents

    private void bamsamImportHelpActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_bamsamImportHelpActionPerformed
        HelpHandler.getHandler().showHelpDialogForKey("rnaseqhelp." + evt.getActionCommand(), main);
    }//GEN-LAST:event_bamsamImportHelpActionPerformed

    private void cancelButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_cancelButtonActionPerformed
        this.setVisible(false);
        cancelled = true;
    }//GEN-LAST:event_cancelButtonActionPerformed

    private void loadGFFButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadGFFButtonActionPerformed
        java.awt.FileDialog pdChooser = new java.awt.FileDialog(new Frame(), "Please choose reference genome annotation GFF3 file", FileDialog.LOAD);
        pdChooser.setDirectory(System.getProperty("user.home"));
        
        System.setProperty("apple.awt.fileDialogForDirectories", "false");
        pdChooser.setVisible(true);
        
        if (pdChooser.getFile() == null) {
            return;
        }
        
        File gff3annoFile = new File(pdChooser.getDirectory(), pdChooser.getFile());
        annoFileField.setText(gff3annoFile.getName());
        
        dataModel.setGFF3annotationFile(gff3annoFile);
        dataModel.setReferenceType(RNASeqDataModel.REFERENCE_TYPE.GENOME);
        
    }//GEN-LAST:event_loadGFFButtonActionPerformed
    
    public SAMFileReader.ValidationStringency getValidationStringency() {
        if ( ((String)validationStringencyBox.getSelectedItem()).equals("strict") ) {
            return SAMFileReader.ValidationStringency.STRICT;
        } else if ( ((String)validationStringencyBox.getSelectedItem()).equals("lenient") ) {
            return SAMFileReader.ValidationStringency.LENIENT;
        } else if ( ((String)validationStringencyBox.getSelectedItem()).equals("silent") ) {
            return SAMFileReader.ValidationStringency.SILENT;
        } else {
            return SAMFileReader.ValidationStringency.DEFAULT_STRINGENCY;
        }
    }
    
    private void importButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_importButtonActionPerformed
        if (genomeButton.isSelected() && annoFileField.getText().equals("")) {
            new SimpleErrorMessage(main, "Please specify a GFF3 annotation file");
            return;
        }  
        
        if (transcriptomeButton.isSelected()) {
            dataModel.setReferenceType(RNASeqDataModel.REFERENCE_TYPE.TRANSCRIPTOME);
        }
        
        setVisible(false);
    }//GEN-LAST:event_importButtonActionPerformed

    private void bamsamImportStringencyHelpActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_bamsamImportStringencyHelpActionPerformed
        HelpHandler.getHandler().showHelpDialogForKey("rnaseqhelp." + evt.getActionCommand(), main);
    }//GEN-LAST:event_bamsamImportStringencyHelpActionPerformed

    private void transcriptomeButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_transcriptomeButtonActionPerformed
        importButton.setEnabled(true);
    }//GEN-LAST:event_transcriptomeButtonActionPerformed

    private void genomeButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_genomeButtonActionPerformed
        importButton.setEnabled(true);
    }//GEN-LAST:event_genomeButtonActionPerformed

    boolean isCancelled() {
        return cancelled;
    }
    
    public void centerOnParent() {
        int x;
        int y;

        // Find out our parent 
        Container myParent = getParent();
        Point topLeft = myParent.getLocationOnScreen();
        Dimension parentSize = myParent.getSize();

        Dimension mySize = getSize();

        if (parentSize.width > mySize.width) {
            x = ((parentSize.width - mySize.width) / 2) + topLeft.x;
        } else {
            x = topLeft.x;
        }

        if (parentSize.height > mySize.height) {
            y = ((parentSize.height - mySize.height) / 2) + topLeft.y;
        } else {
            y = topLeft.y;
        }

        setLocation(x, y);
    }
    /**
     * @param args the commad line arguments
     */
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JTextField annoFileField;
    private javax.swing.JButton bamsamImportHelp;
    private javax.swing.JButton bamsamImportStringencyHelp;
    private javax.swing.JButton cancelButton;
    private javax.swing.JRadioButton genomeButton;
    private javax.swing.JButton importButton;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JLabel jLabel2;
    private javax.swing.JButton loadGFFButton;
    private javax.swing.ButtonGroup refTypeButtonGroup;
    private javax.swing.JRadioButton transcriptomeButton;
    private javax.swing.JComboBox validationStringencyBox;
    private org.jdesktop.beansbinding.BindingGroup bindingGroup;
    // End of variables declaration//GEN-END:variables

    
}
